1951057: Cherry Pick to Add in memory read of cache, delete SCA cert with not needed.

tlhmerry9 opened the pull request:
…eeded

This stops bad cache values from incorrectly displaying that
the system is registered to an organization which is running
in Simple Content Access mode.

Now the organization's content access mode is cached only in memory
so as to limit the duration of it's validity to the lifetime of
any of our applications runtime.

cnsnyder commented:
Thanks for this so far! I think we'll need to pull a few additional commits (or their equivalent changes) from subscription-manager-1.24 branch. The performance feedback we received indicated that it is better to have the status be inaccurate for 4 hours rather than to query the owner once per run. I'll see if I can find the commits and update this with them.

tlhmerry9 commented:
> Thanks for this so far! I think we'll need to pull a few additional commits (or their equivalent changes) from subscription-manager-1.24 branch. The performance feedback we received indicated that it is better to have the status be inaccurate for 4 hours rather than to query the owner once per run. I'll see if I can find the commits and update this with them.

Just let me know - I have a few test to fix in the master branch so I have things to do :)

cnsnyder commented:
I think we need to make sure to clear the cache on refresh (which for subman-1.24 was implemented in this commit 883914c39153f2285852e55d1f2a515d937ddb74).
I'm not sure there is a commit that would apply cleanly to master exactly for the 4 hour timeout but this will be a useful reference if not: 27c07c96184e3a5fdda0bd845a887d6dd4ccc89e

tlhmerry9 commented:
q1

> I think we need to make sure to clear the cache on refresh (which for subman-1.24 was implemented in this commit [883914c](https://github.com/candlepin/subscription-manager/commit/883914c39153f2285852e55d1f2a515d937ddb74)).
> I'm not sure there is a commit that would apply cleanly to master exactly for the 4 hour timeout but this will be a useful reference if not: [27c07c9](https://github.com/candlepin/subscription-manager/commit/27c07c96184e3a5fdda0bd845a887d6dd4ccc89e)

Thank you!

tlhmerry9 commented:
There are 2 commits to represent the 3 commits I cherry-picked from.  I can squash them if prefered.  I am still working on the master branch hoping to finish it today or in the morning.

cnsnyder commented on test/stubs.py:
```diff
@@ -721,7 +721,7 @@ def delete_cache(self):
 
 
-class StubAvailableEntitlementsCache(AvailableEntitlementsCache):
+class StubContentAccessModeCache(ContentAccessModeCache):
```
Any reason to move all these stubs around? It could be an artefact of the cherry-pick but it doesn't seem meaningful to me.

tlhmerry9 replied:
It was an artifact of the cherry-pick, yes.  Doesn't seem meaningful to me either.

tlhmerry9 replied:
I reordered them

cnsnyder commented on test/stubs.py:
```diff
@@ -129,7 +129,7 @@ def stubInitConfig():
 # create a global CFG object,then replace it with our own that candlepin
 # read from a stringio
-#config.get_config_parser(config_file="test/rhsm.conf")
+#config.initConfig(config_file="test/rhsm.conf")
```
Any reason to modify this commented code? Again could be an artefact of the cherry pick but it doesn't seem to be necessary afaict.

tlhmerry9 replied:
it was one of the conflicts when I cherry-picked.  I mistakenly chose to go with what was in 1.24.  Will Change Back.

cnsnyder commented on test/stubs.py:
```diff
@@ -26,8 +26,8 @@
         OverrideStatusCache, ProfileManager, InstalledProductsManager, ReleaseStatusCache, \
-        PoolStatusCache, ContentAccessModeCache, SupportedResourcesCache, AvailableEntitlementsCache, \
-        SyspurposeValidFieldsCache, CurrentOwnerCache
+        PoolStatusCache, ContentAccessModeCache, SupportedResourcesCache, ReadThroughInMemoryCache, \
```
I don't think we should still keep the ReadThroughInMemoryCache (despite how much I liked it) as it wasn't kept in master b/c we needed a different validity period for the ContentAccessModeCache (4 hours rather than once per run).

tlhmerry9 replied:
That was one of the challenges with these cherry-picks - First was implementing ReadThroughInMemoryCache (and everything associated with it) and the next cherry-pick removed it.  I didn't want to remove it from the code until I had confirmation.

cnsnyder commented on src/subscription_manager/utils.py:
```diff
@@ -229,6 +197,32 @@ def get_current_owner(uep=None, identity=None):
     return cache.read_data(uep, identity)
 
 
+def get_content_access_mode(uep=None, identity=None, owner=None):
```
This method is unused and IMO should be removed along with the other InMemory* cache bits.

tlhmerry9 replied:
I agree, am removing.

cnsnyder commented on src/subscription_manager/injectioninit.py:
```diff
@@ -21,8 +21,8 @@
-    SyspurposeComplianceStatusCache, ContentAccessModeCache, \
-    SupportedResourcesCache, AvailableEntitlementsCache, \
+    SyspurposeComplianceStatusCache, SupportedResourcesCache, \
+    AvailableEntitlementsCache, ContentAccessModeCache, \
```
This might be another harmless artefact of the cherry-pick. I'm tempted to say keep it unchanged unless necessary for some reason.

tlhmerry9 replied:
yes, this was a result of the cherry-pick - I am leaving as is.

cnsnyder commented on src/subscription_manager/cache.py:
```diff
@@ -191,6 +195,86 @@ def update_check(self, uep, consumer_uuid, force=False):
             return 0  # No updates performed.
 
 
+class InMemoryCache(object):
```
We should remove this as it is not used in this branch or in the main branch.

tlhmerry9 replied:
removed

cnsnyder commented on src/subscription_manager/cache.py:
```diff
@@ -191,6 +195,86 @@ def update_check(self, uep, consumer_uuid, force=False):
+            return None
+
+
+class ReadThroughInMemoryCache(InMemoryCache):
```
We should remove this b/c it is not used (and/or shouldn't be used as of yet). Please either remove or let me know why we should keep it.

tlhmerry9 replied:
removed

cnsnyder reviewed (changes requested):
Overall on the right track but it looks like some artefacts from the cherry-pick came over and a few from the main branch which are no longer used. Please see the inline comments for details and let me know your thoughts! Thanks for all the help so far!

tlhmerry9 commented:
> Overall on the right track but it looks like some artefacts from the cherry-pick came over and a few from the main branch which are no longer used. Please see the inline comments for details and let me know your thoughts! Thanks for all the help so far!

I agree - I almost removed the InCache* bits as it seemed one cherry-pick implemented them and another removed them.  Wanted to have some clarification, thank you!   I addressed all your comments.

cnsnyder commented on src/subscription_manager/cache.py:
```diff
@@ -20,6 +20,7 @@
 this with the current state, and perform an update on the server if
 necessary.
 """
+import copy
```
This import is breaking the flake8 stylish tests, IMO it should be removed.

tlhmerry9 replied:
It didn't break nosetests for me.  Are you running the tests differently (that way I will know if I need to run additional tests).  But I removed.

cnsnyder commented on test/test_registration.py:
```diff
@@ -84,18 +84,18 @@ def test_registration(self):
-        cmd.main(['register', '--activationkey=test_key', '--org=test_org'])
-        self.mock_register.register.assert_called_once()
-        mock_entcertlib.update.assert_called_once()
+    #@patch('subscription_manager.managercli.EntCertActionInvoker')
```
Why should we comment out this test? Is it not longer necessary? If it is no longer necessary let's remove it entirely. Otherwise we should leave it uncommented.

tlhmerry9 replied:
I broke a few tests (with the same error) so I commented out all but one so I could figure out the fix.  I thought I got them all.
